File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
N, K = map(int, input().split())
dp = [[0] * (K + 1) for _ in range(N + 1)]
things = [[0, 0]]

for _ in range(N):
    weight, value = map(int, input().split())
    things.append([weight, value])

for x in range(1, N + 1):
    w, v = things[x]

    for y in range(1, K + 1):
        if y - w >= 0:
            dp[x][y] = max(dp[x - 1][y], dp[x - 1][y - w] + v)
        else:
            dp[x][y] = dp[x - 1][y]

print(dp[N][K])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.cc
#include <iostream>
#include <algorithm>
using namespace std;

#define max(x, y) (x) > (y) ? (x) : (y);

int bag[101][100001];
int things[101][2];

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int N, K, W, V;
	cin >> N >> K;

	for (int i = 1; i < N + 1; i++)
	{
		cin >> W >> V;
		
		for (int j = 0; j < K + 1; j++)
		{
			if (W > j)	bag[i][j] = bag[i - 1][j];
			else {
				bag[i][j] = max(bag[i - 1][j], bag[i - 1][j - W] + V);
			}
		}
	}

	/*for (int i = 0; i < N + 1; i++)
	{
		for (int j = 0; j < K + 1; j++)
		{
			cout << bag[i][j] << " ";
		}
		cout << "\n";
	}*/

	cout << bag[N][K];
}